src/OrchardCore/OrchardCore/Shell/FeatureProfilesValidationProvider.cs

```diff
@@ -0,0 +1,115 @@
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Options;
+using OrchardCore.Environment.Extensions;
+using OrchardCore.Environment.Shell.Models;
+
+namespace OrchardCore.Environment.Shell
+{
+    public class FeatureProfilesValidationProvider : IFeatureValidationProvider
+    {
+        private readonly IExtensionManager _extensionManager;
+        private readonly FeatureProfilesRuleOptions _featureProfilesRuleOptions;
+        private readonly IShellHost _shellHost;
+        private readonly ShellSettings _shellSettings;
+
+        // Cached across requests as this is called a lot and can be calculated once.
+        private readonly Dictionary<string, bool> _allowed = new Dictionary<string, bool>(StringComparer.OrdinalIgnoreCase);
+        private (bool NotFound, FeatureProfile FeatureProfile) _featureProfileLookup;
+
+        public FeatureProfilesValidationProvider(
+            IExtensionManager extensionManager,
+            IShellHost shellHost,
+            ShellSettings shellSettings,
+            IOptions<FeatureProfilesRuleOptions> featureOptions)
+        {
+            _extensionManager = extensionManager;
+            _shellHost = shellHost;
+            _shellSettings = shellSettings;
+            _featureProfilesRuleOptions = featureOptions.Value;
+        }
+
+        public ValueTask<bool> IsExtensionValidAsync(string id)
+            => IsFeatureValidAsync(id);
+
+        public async ValueTask<bool> IsFeatureValidAsync(string id)
+        {
+            var featureProfileName = _shellSettings["FeatureProfile"];
+            if (String.IsNullOrEmpty(featureProfileName))
+            {
+                return true;
+            }
+
+            if (!_featureProfileLookup.NotFound)
+            {
+                var scope = await _shellHost.GetScopeAsync(ShellHelper.DefaultShellName);
+
+                await scope.UsingAsync(async (scope) =>
+                {
+                    var featureProfilesService = scope.ServiceProvider.GetService<IFeatureProfilesService>();
+
+                    var feauterProfiles = await featureProfilesService.GetFeatureProfilesAsync();
+
+                    if (feauterProfiles.TryGetValue(featureProfileName, out var featureProfile))
+                    {
+                        _featureProfileLookup = (false, featureProfile);
+                    }
+                    else
+                    {
+                        _featureProfileLookup = (true, null);
+                    }
+                });
+            }
+
+            // When the management feature is not enabled we need to pass feature validation.
```

Comment:
@jtkech if you have some time wonder if you could look over this, and the changes I made to the `IShellFeaturesManager`

```
        Task<IEnumerable<IFeatureInfo>> GetAvailableFeaturesAsync();
        Task<IEnumerable<IExtensionInfo>> GetAvailableExtensionsAsync();
```

@sebastienros was keen for this to move to a deeper level than it was, i.e. in the controllers.

Rather than adding functionality to the `IExtensionManager` I've added it to the `IShellFeaturesManager`, as that seemed the appropriate place.

Then I wasn't able to force add a feature to every tenant, so I have had to add it as a shell feature, but through the orchard core builder, and provide a null service, for when the feature is not enabled on the default tenant.

And then use the shell scope to resolve the profiles from the default shell when the service is active.

You might have some better ideas for that?

Reply:
@deanmarcussen No problem, I will take a look tomorrow

Reply:
@deanmarcussen

Yesterday I did a little review but I forgot to submit it ;) Submitted with my new comments

This night I updated VStudio preview (more than one hour with my slow connection) and at this moment I'm not able to build any project, so I didn't have so much time and didn't try yet what i'm suggesting

That said, yes I think we can get rid of extensions so that we just have to define rules for features. The things I would like to retry and confirm is when in a Theme manifest we define multiple features, with dependencies, and so on.

I would say that all its feature could be different themes but in the themes admin page the lookup is currently done at the extensions level. Then looking at some code, e.g. in `ThemeViewLocationExpanderProvider`.

            var currentThemeAndBaseThemesOrdered = _extensionManager
                .GetFeatures(new[] { currentThemeId })
                .Where(x => x.Extension.IsTheme())
                .Reverse()
                .ToList();

Here it looks like that a Theme extension can only define one main feature which is the theme, anyway as said in the theme admin page the lookup is at the extensions level.

That said in the following suggestion, I still assume that a Theme extension may have multiple features, but at least I assume that only one feature is the theme, the feature whose Id is equal to the extension Id (normally the main and then the first feature in the list). So here, one suggestion quickly done in the Themes `AdminController`

This to not have to use `GetAvailableExtensionsAsync()`, only `GetAvailableFeaturesAsync()`

            var themes = (await _shellFeaturesManager.GetAvailableFeaturesAsync())
                .Where(f => f.Extension.IsTheme() && f.Extension.Id == f.Id)
                .Where(f =>
                {
                    var tags = f.Extension.Manifest.Tags.ToArray();
                    var isHidden = tags.Any(t => String.Equals(t, "hidden", StringComparison.OrdinalIgnoreCase));
                    if (isHidden)
                    {
                        return false;
                    }

                    return true;
                })
                .Select(f =>
                {
                    var isAdmin = IsAdminTheme(f.Extension.Manifest);
                    var themeId = isAdmin ? currentAdminTheme?.Extension.Id : currentSiteTheme?.Extension.Id;
                    var isCurrent = f.Extension.Id == themeId;
                    var isEnabled = enabledFeatures.Any(e => e.Id == f.Id);
                    var themeEntry = new ThemeEntry(f.Extension)
                    {
                        Enabled = isEnabled,
                        CanUninstall = installThemes,
                        IsAdmin = isAdmin,
                        IsCurrent = isCurrent
                    };

                    return themeEntry;
                })
                .OrderByDescending(t => t.IsCurrent);

Reply:
Cool, makes sense @jtkech I will check into it and see if we can remove the `GetAvailableExtensionsAsync()`